src/bit-systems/waypoint.ts

```diff
@@ -0,0 +1,209 @@
+import { defineQuery, entityExists, exitQuery, hasComponent } from "bitecs";
+import { Matrix4, Mesh, MeshStandardMaterial, Object3D } from "three";
+import { HubsWorld } from "../app";
+import {
+  HoveredRemoteLeft,
+  HoveredRemoteRight,
+  Interacted,
+  Networked,
+  NetworkedWaypoint,
+  Owned,
+  SceneRoot,
+  Waypoint,
+  WaypointPreview
+} from "../bit-components";
+import { CharacterControllerSystem } from "../systems/character-controller-system";
+import { sleep } from "../utils/async-utils";
+import { anyEntityWith, findAncestorWithComponent } from "../utils/bit-utils";
+import { coroutine } from "../utils/coroutine";
+import { EntityID } from "../utils/networking-types";
+import { takeOwnership } from "../utils/take-ownership";
+import { takeOwnershipWithTime } from "../utils/take-ownership-with-time";
+import { setMatrixWorld } from "../utils/three-utils";
+
+export enum WaypointFlags {
+  canBeSpawnPoint = 1 << 0,
+  canBeOccupied = 1 << 1,
+  canBeClicked = 1 << 2,
+  willDisableMotion = 1 << 3,
+  willDisableTeleporting = 1 << 4,
+  willMaintainInitialOrientation = 1 << 5,
+  snapToNavMesh = 1 << 6
+}
+
+const waypointQuery = defineQuery([Waypoint]);
+
+let myOccupiedWaypoint = 0;
+
+export function releaseOccupiedWaypoint() {
+  if (myOccupiedWaypoint) {
+    NetworkedWaypoint.occupied[myOccupiedWaypoint] = 0;
+    myOccupiedWaypoint = 0;
+  }
+}
+
+function occupyWaypoint(world: HubsWorld, eid: EntityID) {
+  if (!hasComponent(world, Owned, eid)) {
+    throw new Error("Tried to occupy waypoint before owning it.");
+  }
+  NetworkedWaypoint.occupied[eid] = 1;
+  myOccupiedWaypoint = eid;
+}
+
+function nonOccupiableSpawnPoints(world: HubsWorld) {
+  return waypointQuery(world).filter(eid => {
+    const canBeSpawnPoint = Waypoint.flags[eid] & WaypointFlags.canBeSpawnPoint;
+    const canBeOccupied = Waypoint.flags[eid] & WaypointFlags.canBeOccupied;
+    return canBeSpawnPoint && !canBeOccupied && findAncestorWithComponent(world, SceneRoot, eid);
+  });
+}
+
+function occupiableSpawnPoints(world: HubsWorld) {
+  return waypointQuery(world).filter(eid => {
+    const canBeSpawnPoint = Waypoint.flags[eid] & WaypointFlags.canBeSpawnPoint;
+    const canBeOccupied = Waypoint.flags[eid] & WaypointFlags.canBeOccupied;
+    return (
+      canBeSpawnPoint &&
+      canBeOccupied &&
+      !NetworkedWaypoint.occupied[eid] &&
+      findAncestorWithComponent(world, SceneRoot, eid)
+    );
+  });
+}
+
+function* tryOccupyAndSpawn(world: HubsWorld, characterController: CharacterControllerSystem, spawnPoint: EntityID) {
+  moveToWaypoint(world, spawnPoint, characterController, true);
+  takeOwnershipWithTime(world, spawnPoint, 0);
```

Comment:
It seem this would fail if the waypoint was ever owned by anyone before us. 

We didn't think we could do a regular `takeOwnership` here, because we didn't want to assume that we'll have already received netcode messages about this entity. (We wanted to take "soft" ownership, instead.)

It would be nice for us to avoid having to add code here (and for networked videos, for example) that said, "Wait long enough to be sure that no one is trying to tell us about their ownership over this entity so we don't accidentally takeover that state." But in this case, the timestamp `0` doesn't seem correct.

Maybe this should be `takeOwnershipWithTime(world, spawnPoint, Networked.lastOwnerTime[spawnPoint] + 1)` , which will be `1` if we've never received a message about this entity. If it's not `1`, then it's as if we're saying, "I want to take ownership and believe I can according to the information I've received as of this timestamp."

Reply:
(This `lastOwnerTime + 1` is also what we do when taking ownership of the entities that are left behind by a parting client.)

Reply:
I wonder if this is actually what we always want to do in spots where we were using 0. We should talk through it.